src/bootstrap/bootstrap.py

```diff
@@ -0,0 +1,326 @@
+            filename = "cargo-nightly-" + self.build + ".tar.gz"
+            url = "https://static.rust-lang.org/cargo-dist/" + self.snap_cargo_date()
+            tarball = os.path.join(cargo_cache, filename)
+            if not os.path.exists(tarball):
+                get(url + "/" + filename, tarball,
+                             quiet=not self.verbose)
+            unpack(tarball, self.bin_root(), filter="cargo",
+                            quiet=not self.verbose)
+            with open(self.cargo_stamp(), 'w') as f:
+                f.write(self.snap_cargo_date())
+
+    def snap_cargo_date(self):
+        if not hasattr(self, '_cargo_date'):
+            self.parse_nightly_dates()
+        return self._cargo_date
+
+    def snap_rustc_date(self):
+        if not hasattr(self, '_rustc_date'):
+            self.parse_nightly_dates()
+        return self._rustc_date
+
+    def rustc_stamp(self):
+        return os.path.join(self.bin_root(), '.rustc-stamp')
+
+    def cargo_stamp(self):
+        return os.path.join(self.bin_root(), '.cargo-stamp')
+
+    def rustc_out_of_date(self):
+        if not os.path.exists(self.rustc_stamp()):
+            return True
+        with open(self.rustc_stamp(), 'r') as f:
+            return self.snap_rustc_date() != f.read()
+
+    def cargo_out_of_date(self):
+        if not os.path.exists(self.cargo_stamp()):
+            return True
+        with open(self.cargo_stamp(), 'r') as f:
+            return self.snap_cargo_date() != f.read()
+
+    def bin_root(self):
+        return os.path.join(self.build_dir, self.build, "stage0")
+
+    def get_toml(self, key):
+        for line in iter(self.config_toml.splitlines()):
+            if line.startswith(key + ' ='):
+                return self.get_string(line)
+        return None
+
+    def get_mk(self, key):
+        for line in iter(self.config_mk.splitlines()):
+            if line.startswith(key):
+                return line[line.find(':=') + 2:].strip()
+        return None
+
+    def cargo(self):
+        config = self.get_toml('cargo')
+        if config:
+            return config
+        return os.path.join(self.bin_root(), "bin/cargo" + self.exe_suffix())
+
+    def rustc(self):
+        config = self.get_toml('rustc')
+        if config:
+            return config
+        config = self.get_mk('CFG_LOCAL_RUST')
+        if config:
+            return config + '/bin/rustc' + self.exe_suffix()
+        return os.path.join(self.bin_root(), "bin/rustc" + self.exe_suffix())
+
+    def get_string(self, line):
+        start = line.find('"')
+        end = start + 1 + line[start+1:].find('"')
+        return line[start+1:end]
+
+    def exe_suffix(self):
+        if sys.platform == 'win32':
+            return '.exe'
+        else:
+            return ''
+
+    def parse_nightly_dates(self):
+        nightlies = os.path.join(self.rust_root, "src/nightlies.txt")
+        rustc, cargo = open(nightlies, 'r').read().split("\n")[:2]
```

Comment:
Might be good to use a context manager here (`with open(..):`) so the file gets properly closed